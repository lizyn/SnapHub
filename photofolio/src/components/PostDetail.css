.post-modal-main {
    display: flex;
    align-items: center;
    flex-direction: column;
    justify-content: center;
    margin-top: 10rem;
}

.post-detail-left {
    display: flex;
    flex-direction: column;
    width: 94rem;
    height: 100rem;
}

.post-detail-image {
    width: 100%;
    height: 30%;
    object-fit: contain;
    border-top-left-radius: 10px;
    background-color:black;
}
.post-detail-description{
    display: flex;
    flex-direction: column;
    gap:0 !important
}
.post-detail-description-user{
    margin-left: 6.5rem;
    margin-bottom: 0 !important;
    font-family: Verdana;
    font-size: 16px;
    font-weight: 600;
    text-transform: uppercase;
}

.post-detail-description-text{
    margin-left: 6.5rem;
    height: 2rem;
    width: 80%;
    overflow-y:scroll;
}
.post-detail-description-text p{
    font-family: Verdana;
    font-size: 12px;
}

.post-detail-right {
    display: flex;
    flex-direction: column;
    width: 100%;
}

.post-detail-actions{
    border-top: 1px solid lightgrey;
    padding: 0.92rem 0.8rem 1rem 0;
    display: flex;
    flex-direction: column;
}

.post-detail-postStats{
    display: flex;
    flex-direction: row;
    margin-bottom: 1em;
}

.post-detail-stats{
    display: flex;
    flex-direction: row;
    height: 2em;
    margin-left: 1.1em;
    width:fit-content;
    padding-right: 1em;
    justify-content: flex-start;
    align-items: center;
    column-gap: 0.5em
}

.MuiModal-root {
    z-index:3000 !important;
}

.post-detail-stats p{
    font-size: 15px;
    text-align: center;
} 

.post-detail-comments {
    height: 80%;
    overflow-y: scroll;
    overflow-x:hidden;
    overflow-wrap: break-word;
    /* background-color:aqua; */
    padding-right: 1rem;
}

.post-detail-post-comment {
    display: flex;
    justify-content: center;
    width: 100%;
    margin-bottom: 0.5rem;
}

.post-detail-post-comment input{
    border-radius: 10px;
    border: 1px solid lightslategrey;
    padding: 0.5em;
    width: 74%;
    overflow-x: scroll;
}

.post-detail-post-comment button{
    border-radius: 8px;
    border: 1px solid lightslategrey;
    width: 4em;
    height: 2.5em;
    margin: 0 5px 0 0;
    background-color:white;
}

.post-detail-post-comment button:hover{
    background-color: rgb(237, 237, 237);
    transition-duration: 200ms;
}

.post-detail-post-comment button:active{
    background-color: rgb(170, 170, 170);
}

.post-detail-post-comment button img{
    width: 1.5em;
    filter: brightness(4);
}

.post-detail-userhead {
    margin: 1rem 0 0 0.5rem;
    padding-bottom: 1rem;
    border-bottom: 1px solid lightgrey;
}
