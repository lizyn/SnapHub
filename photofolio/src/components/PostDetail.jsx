import React, { useEffect, useState } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import Box from '@mui/material/Box';
import Modal from '@mui/material/Modal';
import { Avatar } from '@mui/material';
import { Link } from 'react-router-dom';
import IconButton from '@mui/material/IconButton';
import LikeIconOutlined from '@mui/icons-material/ThumbUpOutlined';
import LikeIconFilled from '@mui/icons-material/ThumbUp';
import { MentionsInput, Mention } from 'react-mentions';
// import { useHistory } from 'react-router-dom';
import { rootUrl } from './Config';
import commentIcon from '../icons/Comment.svg';
import sendIcon from '../icons/Send.svg';
import {
  fetchComments,
  likePosts,
  createComment,
  deleteComment,
  deletePost
} from '../api/axios';
import CommentRow from './CommentRow';
import './PostDetail.css';
import EditPostModals from './EditPostModal';

const style = {
  position: 'absolute',
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'flex-start',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)',
  width: 1000,
  maxHeight: 600,
  bgcolor: 'background.paper',
  borderRadius: '10px',
  boxShadow: 24,
  zIndex: 4000
};

function PostDetail(props) {
  PostDetail.propTypes = {
    open: PropTypes.bool.isRequired,
    setOpen: PropTypes.func.isRequired,
    author: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    postTimeStr: PropTypes.string,
    img: PropTypes.string,
    avatar: PropTypes.string,
    likes: PropTypes.number,
    postId: PropTypes.string.isRequired,
    userId: PropTypes.string.isRequired,
    handlePostChange: PropTypes.func.isRequired
  };

  PostDetail.defaultProps = {
    postTimeStr: 'a while ago',
    img: '/',
    avatar: '/',
    likes: 0
  };

  const {
    open,
    setOpen,
    img,
    author,
    avatar,
    likes,
    title,
    postTimeStr,
    postId,
    userId,
    handlePostChange
  } = props;

  const [comments, setComments] = useState([]);
  const [postLiked, setPostLiked] = useState(false);
  const [commentInput, setCommentInput] = useState('');
  const [commentSubmit, setCommentSubmit] = useState('');
  const [mentionCandidates, setMentionCandidates] = useState([]);

  const getMentionCandidates = async () => {
    try {
      const response = await axios.get(`${rootUrl}/users`);
      // console.log(response);
      const candidates = response.data.data.map((user) => ({
        // eslint-disable-next-line no-underscore-dangle
        id: user._id,
        display: `${user.firstName} ${user.lastName}`
      }));
      // console.log('----------');
      // console.log(candidates);
      setMentionCandidates(candidates);
    } catch (err) {
      console.log(err);
      // return err;
    }
  };
  const [postDeleted, setPostDeleted] = useState(false);
  const [commentEdited, setCommentEdited] = useState(false);
  // const [postModalIsOpen, setPostModalOpen] = useState(false);
  const [testState, setTestState] = useState(false);
  // const closePostModal = () => setPostModalOpen(false);
  const closeEditPostModal = () => setTestState(false);
  let commentlist;
  // const [alert, setAlert] = useState(false);
  // const history = useHistory();

  useEffect(() => {
    async function fetchData() {
      const commentsData = await fetchComments(postId);
      if (Array.isArray(commentsData) && commentsData.length >= 0) {
        commentlist = commentsData;
        setComments(commentlist);
      }
    }
    fetchData();
  }, [commentSubmit, postDeleted, commentEdited]);

  useEffect(() => {
    getMentionCandidates();
  }, []);

  const handleClose = (e, r) => {
    if (r === 'backdropClick') {
      setOpen(false);
    }
  };

  const handleLikeClick = () => {
    setPostLiked((currentLike) => !currentLike);
    if (!postLiked) {
      likePosts(postId, likes + 1);
    } else {
      likePosts(postId, likes);
    }
  };

  const handlePostEdit = async () => {
    setOpen(false);
    setTestState((x) => !x);
  };

  const handlePostDelete = async (id) => {
    try {
      const response = await deletePost(id);
      handlePostChange(id);
      return response.status;
    } catch (err) {
      return err.message;
    }
  };

  const handleCommentDelete = async (commentId) => {
    await deleteComment(commentId);
    setPostDeleted((currentDelete) => !currentDelete);
  };

  const populateComments = () => {
    const allComments = [];
    comments.forEach((comment) => {
      allComments.push(
        <CommentRow
          // eslint-disable-next-line no-underscore-dangle
          key={comment._id}
          userId={comment.userId}
          commentText={comment.text}
          // eslint-disable-next-line no-underscore-dangle
          commentId={comment._id}
          commentDel={handleCommentDelete}
          commentEd={setCommentEdited}
        >
          {comment.text}
        </CommentRow>
      );
    });
    return allComments;
  };

  const handleCommentChange = (e) => {
    setCommentInput(e.target.value);
  };

  const convertMentionInComment = (comment) => {
    let clickableComment = comment;
    clickableComment = clickableComment
      .split('@@@^')
      .join('<a href="/profile/');
    clickableComment = clickableComment.split('@@@|').join('">@');
    clickableComment = clickableComment.split('$@@@').join('</a>');
    clickableComment = `${clickableComment}`;
    return clickableComment;
  };

  const handleCommentSubmit = async () => {
    const comment = convertMentionInComment(commentInput);
    await createComment('63899e8d4bd2e0bd159d0e10', postId, comment);
    setCommentSubmit(comment);
    setCommentInput('');
  };

  const allComments = populateComments();
  return (
    <div className="post-modal-main">
      <EditPostModals
        closeModal={closeEditPostModal}
        open={testState}
        setAlert={() => {}}
        postId={postId}
        title={title}
        img={img}
      />
      <Modal
        open={open || false}
        onClose={handleClose}
        aria-labelledby="post-title"
        aria-describedby="post-description"
      >
        <Box className="post-detail-text" sx={style}>
          <div className="post-detail-left">
            <img src={img} className="post-detail-image" alt="post" />
            <div className="post-detail-description">
              <Link to={`/profile/${userId}`}>
                <Avatar
                  alt="me"
                  className="Avatar"
                  src={avatar}
                  sx={{
                    width: 70,
                    height: 70,
                    position: 'absolute',
                    top: '77%',
                    left: '2%'
                  }}
                />
              </Link>
              <p className="post-detail-description-user">{title}</p>
              <div className="post-detail-description-text">
                <p>text description</p>
              </div>
            </div>
          </div>
          <div className="post-detail-right">
            <div className="post-detail-userhead">
              <Link to={`/profile/${userId}`}>
                <Avatar
                  alt="me"
                  className="Avatar"
                  src={avatar}
                  sx={{ width: 60, height: 60 }}
                />
              </Link>
              <p className="postUsername">{author}</p>
              <p className="postTime">{postTimeStr}</p>
            </div>
            <div>
              <button type="submit" onClick={() => handlePostEdit(postId)}>
                Edit Post
              </button>
            </div>
            <div>
              <button
                type="submit"
                className="deleteButton"
                onClick={() => handlePostDelete(postId)}
              >
                Delete Post
              </button>
            </div>
            <div className="post-detail-comments">{allComments}</div>
            <div className="post-detail-actions">
              <div className="post-detail-postStats">
                <div className="post-detail-stats">
                  <IconButton
                    onClick={handleLikeClick}
                    aria-label="like"
                    sx={{ curser: 'pointer' }}
                  >
                    {postLiked ? (
                      <LikeIconFilled
                        sx={{
                          color: 'orange!important'
                        }}
                      />
                    ) : (
                      <LikeIconOutlined />
                    )}
                  </IconButton>
                  <p>{postLiked ? `${likes + 1} Likes` : `${likes} Likes`}</p>
                </div>
                <div className="post-detail-stats">
                  <img src={commentIcon} alt="comment" />
                  <p>{comments.length} Comments</p>
                </div>
              </div>
              <div className="post-detail-post-comment">
                <button type="submit" onClick={handleCommentSubmit}>
                  <img src={sendIcon} alt="send comment" />
                </button>
                <MentionsInput
                  className="comments-textarea"
                  value={commentInput}
                  placeholder="Post a comment. Mention with '@'"
                  onChange={handleCommentChange}
                  singleLine
                  allowSpaceInQuery
                  style={{
                    width: '75%',
                    maxWidth: '20em'
                  }}
                >
                  <Mention
                    trigger="@"
                    data={mentionCandidates}
                    markup="@@@^__id__@@@|__display__$@@@"
                    style={{ backgroundColor: '#FFDD2B' }}
                    displayTransform={(id, display) => `@${display}`}
                    // renderSuggestion={this.renderUserSuggestion}
                  />
                </MentionsInput>
              </div>
            </div>
          </div>
        </Box>
      </Modal>
    </div>
  );
}

export default PostDetail;
