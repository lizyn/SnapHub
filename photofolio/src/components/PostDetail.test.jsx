/**
 * @jest-environment jsdom
 */

import { React } from 'react';
import { render, screen } from '@testing-library/react';
// import { renderIntoDocument, fireEvent } from 'react-testing-library';
import '@testing-library/jest-dom/extend-expect';
// import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import PostDetail from './PostDetail';
// import axios from '../api/axios';

// snapshot testing
test('Post Detail matches snapshot', () => {
  const component = renderer.create(<PostDetail />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('render 4 buttons', () => {
  render(
    <PostDetail
      open
      setOpen={() => {}}
      author="Test"
      title=""
      img=""
      avatar=""
      likes={1}
      postId={2}
    />
  );
  // const buttons = screen.getAllByRole('button');
  // arrow button, upload a file button, close button, post button
  // expect(buttons.length).toBe(4);
});
