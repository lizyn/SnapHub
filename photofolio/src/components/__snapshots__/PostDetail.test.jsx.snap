// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Detail matches snapshot 1`] = `
<div
  className="post-modal-main"
/>
`;
